### PR TITLE
Optimize outputs accumulating for SegmentTermsEnum and IntersectTermsEnum 

### PR DESCRIPTION
**Pull request by gf2121**
## Description
Previous talk is too long to track so i opened a new PR to make a summery here. More details are available in https://github.com/apache/lucene/pull/12661.

After merging of https://github.com/apache/lucene/pull/12631, we found a [regression](https://home.apache.org/~mikemccand/lucenebench/2023.10.10.18.03.55.html) for term-dictionary-related tasks like PKLookup, Fuzzy ... After some digging I suspect that the regression is caused by  more `Outputs#add` and `Outputs#read` on reading side, as the 'MSB VLong output format' making FST sharing more output prefix. The difference of calling times of `Outputs#add` and `Outputs#read` is shown below:

  | LSB VLong | MSB VLong | diff
-- | -- | -- | --
Outputs#read times | 116097 | 149803 | 29.03%
Outputs#add times | 144 | 111568 | 77377.78%

## Solution

This patch tries to do two optimizations to get back the speed:
* Instead of combining all outputs into single one, this patch collect all BytesRefs into an array and build a DataInput view over it, reducing the cost of objects construction and memory copy.
* Instead of copying bytes into floor data, this patch directly uses the last output as floor data. Floor data is guaranteed continuous in single arc because we can not have two same `fp` encoded before floor data.

## Benchmark

> Both run 30 JVM round and 50 tasks per JVM:

### Comparing to after merging of https://github.com/apache/lucene/pull/12631

BaseLine ( **After** merging of https://github.com/apache/lucene/pull/12631 )

> $ git log
> Write MSB VLong for better outputs sharing in block tree index (#12631)
> DeletedTerms#clear should reset ByteBlockPool (#12630)

Candidate

> $ git log
> Optimize output accumulating
> Write MSB VLong for better outputs sharing in block tree index
> DeletedTerms#clear should reset ByteBlockPool

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
                   OrHighNotHigh      166.79      (4.5%)      165.12      (4.3%)   -1.0% (  -9% -    8%) 0.376
                    OrHighNotLow      258.47      (4.3%)      256.46      (3.3%)   -0.8% (  -7% -    7%) 0.430
            HighIntervalsOrdered        0.38      (3.3%)        0.38      (3.1%)   -0.8% (  -6% -    5%) 0.359
                 LowSloppyPhrase        1.64      (3.4%)        1.63      (4.9%)   -0.7% (  -8% -    7%) 0.544
                   OrNotHighHigh      150.83      (5.1%)      149.86      (4.4%)   -0.6% (  -9% -    9%) 0.597
                HighSloppyPhrase        3.67      (3.3%)        3.64      (3.6%)   -0.6% (  -7% -    6%) 0.470
                         Prefix3       67.16      (4.3%)       66.78      (2.8%)   -0.6% (  -7% -    6%) 0.543
                    OrNotHighMed      135.65      (4.4%)      134.95      (3.7%)   -0.5% (  -8% -    7%) 0.620
                    OrNotHighLow      257.14      (2.0%)      255.85      (1.7%)   -0.5% (  -4% -    3%) 0.291
                     AndHighHigh       19.33      (2.8%)       19.25      (2.2%)   -0.4% (  -5% -    4%) 0.517
                    OrHighNotMed      128.51      (5.3%)      128.06      (4.2%)   -0.3% (  -9% -    9%) 0.778
                 MedSloppyPhrase        9.56      (3.6%)        9.54      (3.8%)   -0.3% (  -7% -    7%) 0.788
                      OrHighHigh       13.23      (3.3%)       13.20      (3.2%)   -0.2% (  -6% -    6%) 0.801
                       OrHighLow      199.40      (1.9%)      199.18      (1.7%)   -0.1% (  -3% -    3%) 0.812
                        Wildcard       69.22      (2.7%)       69.19      (1.9%)   -0.0% (  -4% -    4%) 0.952
                      AndHighMed       42.14      (3.7%)       42.14      (2.9%)    0.0% (  -6% -    6%) 0.993
                    HighSpanNear        7.00      (3.1%)        7.00      (3.1%)    0.1% (  -5% -    6%) 0.945
                          Fuzzy1       44.05      (1.8%)       44.08      (1.7%)    0.1% (  -3% -    3%) 0.872
                       OrHighMed       38.54      (2.6%)       38.59      (3.2%)    0.1% (  -5% -    6%) 0.871
                         Respell       28.40      (2.2%)       28.45      (2.2%)    0.2% (  -4% -    4%) 0.790
                     LowSpanNear        4.23      (1.7%)        4.24      (1.5%)    0.2% (  -2% -    3%) 0.580
             MedIntervalsOrdered        3.47      (6.6%)        3.49      (7.3%)    0.3% ( -12% -   15%) 0.856
             LowIntervalsOrdered        2.76      (4.3%)        2.77      (5.0%)    0.4% (  -8% -   10%) 0.761
                          Fuzzy2       35.28      (1.6%)       35.42      (1.6%)    0.4% (  -2% -    3%) 0.325
                      HighPhrase       38.91      (2.6%)       39.07      (2.5%)    0.4% (  -4% -    5%) 0.532
                         MedTerm      248.39      (4.2%)      249.48      (4.2%)    0.4% (  -7% -    9%) 0.685
                       LowPhrase       26.44      (2.0%)       26.57      (2.3%)    0.5% (  -3% -    4%) 0.385
                       MedPhrase        9.51      (2.3%)        9.57      (2.6%)    0.6% (  -4% -    5%) 0.365
                      AndHighLow      338.26      (3.0%)      340.61      (2.6%)    0.7% (  -4% -    6%) 0.339
                         LowTerm      242.60      (4.5%)      244.38      (4.8%)    0.7% (  -8% -   10%) 0.545
                     MedSpanNear        1.11      (1.4%)        1.12      (1.5%)    0.8% (  -2% -    3%) 0.042
                        HighTerm      230.63      (3.8%)      233.01      (4.6%)    1.0% (  -7% -    9%) 0.339
                        PKLookup      100.76      (2.9%)      105.17      (3.2%)    4.4% (  -1% -   10%) 0.000
```

### Comparing to before merging of https://github.com/apache/lucene/pull/12631

BaseLine ( **Before** merging of https://github.com/apache/lucene/pull/12631 )

> $ git log
> DeletedTerms#clear should reset ByteBlockPool 

Candidate

> $ git log
> Optimize output accumulating
> Write MSB VLong for better outputs sharing in block tree index
> DeletedTerms#clear should reset ByteBlockPool

(Not sure why seeing tiny speed up for Orxxx tasks. I suspect it is a noise)
```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
             MedIntervalsOrdered        3.50      (7.9%)        3.43      (7.4%)   -2.2% ( -16% -   14%) 0.269
                        Wildcard       70.27      (2.9%)       69.22      (2.2%)   -1.5% (  -6% -    3%) 0.023
            HighIntervalsOrdered        0.38      (4.0%)        0.37      (4.0%)   -1.3% (  -8% -    6%) 0.194
             LowIntervalsOrdered        2.78      (5.3%)        2.75      (5.2%)   -1.1% ( -10% -    9%) 0.417
                         Prefix3       66.95      (4.3%)       66.25      (3.1%)   -1.0% (  -8% -    6%) 0.282
                         LowTerm      246.48      (8.3%)      244.15      (6.8%)   -0.9% ( -14% -   15%) 0.631
                       MedPhrase        9.57      (3.6%)        9.50      (3.1%)   -0.8% (  -7% -    6%) 0.366
                HighSloppyPhrase        3.63      (3.5%)        3.62      (3.8%)   -0.3% (  -7% -    7%) 0.761
                      HighPhrase       39.07      (3.5%)       38.96      (3.3%)   -0.3% (  -6% -    6%) 0.749
                        PKLookup      105.27      (2.8%)      105.13      (3.0%)   -0.1% (  -5% -    5%) 0.859
                 MedSloppyPhrase        9.49      (3.2%)        9.48      (3.9%)   -0.1% (  -6% -    7%) 0.920
                         Respell       28.45      (2.6%)       28.46      (2.2%)    0.0% (  -4% -    5%) 0.957
                 LowSloppyPhrase        1.63      (5.0%)        1.63      (5.6%)    0.1% (  -9% -   11%) 0.929
                       LowPhrase       26.31      (3.3%)       26.36      (2.1%)    0.2% (  -5% -    5%) 0.782
                          Fuzzy1       43.71      (1.3%)       43.82      (1.7%)    0.3% (  -2% -    3%) 0.509
                   OrHighNotHigh      164.97      (3.7%)      165.40      (4.4%)    0.3% (  -7% -    8%) 0.806
                     AndHighHigh       19.17      (2.1%)       19.23      (2.3%)    0.3% (  -3% -    4%) 0.569
                     MedSpanNear        1.11      (2.0%)        1.12      (2.0%)    0.6% (  -3% -    4%) 0.266
                      AndHighLow      338.06      (2.1%)      340.30      (1.9%)    0.7% (  -3% -    4%) 0.196
                     LowSpanNear        4.20      (2.4%)        4.23      (2.1%)    0.7% (  -3% -    5%) 0.194
                      OrHighHigh       13.15      (4.0%)       13.26      (4.5%)    0.8% (  -7% -    9%) 0.454
                   OrNotHighHigh      148.59      (3.9%)      149.94      (5.1%)    0.9% (  -7% -   10%) 0.442
                        HighTerm      229.24      (3.7%)      231.60      (4.2%)    1.0% (  -6% -    9%) 0.314
                    HighSpanNear        6.94      (3.3%)        7.01      (3.6%)    1.1% (  -5% -    8%) 0.224
                    OrHighNotMed      126.33      (4.3%)      127.74      (4.9%)    1.1% (  -7% -   10%) 0.343
                    OrHighNotLow      254.19      (3.1%)      257.19      (3.9%)    1.2% (  -5% -    8%) 0.200
                      AndHighMed       41.35      (5.0%)       42.00      (3.2%)    1.6% (  -6% -   10%) 0.148
                          Fuzzy2       34.61      (1.8%)       35.34      (2.1%)    2.1% (  -1% -    6%) 0.000
                       OrHighMed       37.84      (5.8%)       38.65      (4.2%)    2.1% (  -7% -   12%) 0.107
                    OrNotHighLow      248.52      (2.8%)      254.75      (2.5%)    2.5% (  -2% -    7%) 0.000
                       OrHighLow      193.81      (2.3%)      200.07      (2.5%)    3.2% (  -1% -    8%) 0.000
                    OrNotHighMed      130.29      (3.6%)      134.66      (4.7%)    3.4% (  -4% -   12%) 0.002
                         MedTerm      241.15      (3.8%)      249.46      (4.5%)    3.4% (  -4% -   12%) 0.001
```

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/FieldReader.java [hunk 1]
```diff
@@ -118,13 +118,11 @@ long readVLongOutput(DataInput in) throws IOException {
    * <p>Package private for testing.
    */
   static long readMSBVLong(DataInput in) throws IOException {
-    long l = 0L;
-    while (true) {
-      byte b = in.readByte();
+    byte b = in.readByte();
+    long l = b & 0x7FL;
+    while (b < 0) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/FieldReader.java:123]: Why this change?  Is it a bit of specialization so CPU can predict branch a bit better?

**Review by mikemccand — COMMENTED** [review]
Net/net I love this idea -- it's basically the same idea as `StringBuilder` (append small strings to list and concatenate only in the end), better than the O(N^2) "just keep concatenating a single bigger and bigger string" that FST does today.

I love that this change erases a longstanding `TODO` also!

I wonder if we could make this appendable output more widely accessible (in `oal.fst.util`) for other use cases that need `Outputs<BytesRef>`, but let's not do that here.

I left some small comments ... I think this is close.  It would be nice to recover the `PKLookup` perf hit.  Thanks @gf2121!

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/FieldReader.java (hunk 1 above)
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/FieldReader.java:123]: Background: https://github.com/apache/lucene/issues/12659#issuecomment-1759244758

This was a try to get back the PKLookUp speed but Luceneutil has shown that this change does not affect performance. I kept it here because I think this way is more consistent with the `DataInput#readVInt` and more readable. But now I would revert it since it is not really related to the point of this PR.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java [hunk 2]
```diff
@@ -142,12 +138,20 @@ public void setState(int state) {
   }
 
   void load(BytesRef frameIndexData) throws IOException {
-    if (frameIndexData != null) {
```
[the remark this answers is not in the document]
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java:145]: Yes. We checked null before for [line 119](https://github.com/apache/lucene/pull/12699/files/71ef88acf5bc3b2c47086b55d63d64720b4232b7#diff-77d427360dcf008af183201e0e748d77ccadae1b2359fa96940c407b48b3f85bR119). It is cast to long now and we do the null check in [line 153](https://github.com/apache/lucene/pull/12699/files/71ef88acf5bc3b2c47086b55d63d64720b4232b7#diff-77d427360dcf008af183201e0e748d77ccadae1b2359fa96940c407b48b3f85bR153) instead.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java [hunk 3]
```diff
@@ -1190,4 +1176,65 @@ public void seekExact(long ord) {
   public long ord() {
     throw new UnsupportedOperationException();
   }
+
+  static class OutputAccumulator extends DataInput {
+
+    /** We could have at most 10 no-empty arcs: 9 for vLong and 1 for floor data. */
+    private static final int MAX_ARC = 10;
```
[the remark this answers is not in the document]
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java]: +1, growing on demand is much more flexible!

**Comment by mikemccand** [comment]
I'll try to review the latest PR soon -- thanks @gf2121.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java [hunk 4]
```diff
@@ -142,12 +138,20 @@ public void setState(int state) {
...
-      // flag, isFloor flag
-      final long code = ite.fr.readVLongOutput(floorDataReader);
-      if ((code & Lucene90BlockTreeTermsReader.OUTPUT_FLAG_IS_FLOOR) != 0) {
+    floorDataReader.reset(frameIndexData.bytes, frameIndexData.offset, frameIndexData.length);
+    load(ite.fr.readVLongOutput(floorDataReader));
+  }
+
+  void load(SegmentTermsEnum.OutputAccumulator outputAccumulator) throws IOException {
+    outputAccumulator.prepareRead();
+    final long code = ite.fr.readVLongOutput(outputAccumulator);
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java:147]: Remove this silly `final`?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java [hunk 5]
```diff
@@ -142,12 +138,20 @@ public void setState(int state) {
...
+    outputAccumulator.setFloorData(floorDataReader);
+    load(code);
+  }
+
+  void load(Long boxCode) throws IOException {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java:152]: Hmm rename to `blockCode`?  Not sure why it's a box :)

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java [hunk 6]
```diff
@@ -142,12 +138,20 @@ public void setState(int state) {
...
+    if (boxCode != null) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java:153]: Maybe add comment that this means this block is the first one in a possible sequence of floor blocks corresponding to a single seek point from the FST terms index?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java [hunk 7]
```diff
@@ -373,9 +380,6 @@ public boolean seekExact(BytesRef target) throws IOException {
 
       int cmp = 0;
 
-      // TODO: reverse vLong byte order for better FST
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:376]: Yay, stale `TODO` garbage collection!

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java [hunk 8]
```diff
@@ -1190,4 +1176,63 @@ public void seekExact(long ord) {
...
+  static class OutputAccumulator extends DataInput {
+
+    BytesRef[] outputs = new BytesRef[16];
+    BytesRef current;
+    int num;
+    int outputIndex;
+    int index;
+
+    void push(BytesRef output) {
+      if (output != Lucene90BlockTreeTermsReader.NO_OUTPUT) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1189]: Might be better to just check `if (output.length > 0) {`? It'd handle cases (not sure this happens) where caller is passing an empty `BytesRef` that is not `NO_OUTPUT`?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java [hunk 9]
```diff
@@ -1190,4 +1176,63 @@ public void seekExact(long ord) {
...
+    }
+
+    void prepareRead() {
+      this.index = 0;
+      this.outputIndex = 0;
+      this.current = outputs[0];
+    }
+
+    void setFloorData(ByteArrayDataInput floorData) {
+      assert outputIndex == num - 1
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1210]: Could we also `assert index == 0` in this assert?  I.e. we are at the start of the last `BytesRef`?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java [hunk 10]
```diff
@@ -1190,4 +1176,63 @@ public void seekExact(long ord) {
...
+      this.num = 0;
+    }
+
+    void prepareRead() {
+      this.index = 0;
+      this.outputIndex = 0;
+      this.current = outputs[0];
+    }
+
+    void setFloorData(ByteArrayDataInput floorData) {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1209]: Could you add a comment that, while this reads/decodes bytes into `floorData`, it does not alter the read position?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java [hunk 11]
```diff
@@ -104,13 +104,9 @@ public SegmentTermsEnumFrame(SegmentTermsEnum ste, int ord) throws IOException {
...
-  public void setFloorData(ByteArrayDataInput in, BytesRef source) {
-    final int numBytes = source.length - (in.getPosition() - source.offset);
-    if (numBytes > floorData.length) {
-      floorData = new byte[ArrayUtil.oversize(numBytes, 1)];
-    }
-    System.arraycopy(source.bytes, source.offset + in.getPosition(), floorData, 0, numBytes);
-    floorDataReader.reset(floorData, 0, numBytes);
+  public void setFloorData(SegmentTermsEnum.OutputAccumulator outputAccumulator) {
+    outputAccumulator.setFloorData(floorDataReader);
+    rewindPos = floorDataReader.getPosition();
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java:109]: Can we move this above line 107?  It smells a bit putting this after -- it requires knowing `setFloorData` doesn't change the read position.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java [hunk 12]
```diff
@@ -247,7 +243,7 @@ void rewind() {
     nextEnt = -1;
     hasTerms = hasTermsOrig;
     if (isFloor) {
-      floorDataReader.rewind();
+      floorDataReader.setPosition(rewindPos);
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java:246]: I wonder if this is still needed?  `setFloorData` didn't change the position?  Or might other operations before the `reset()` have altered the read position?  Hard to think about.

**Review by mikemccand — COMMENTED** [review]
Thanks @gf2121 -- I left mostly smallish comments, but some head scratching -- hard to remember how these two crazy iterators work.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java (hunk 8 above)
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1189]: This object comparing is what we did before. I believe it is right because we have strict [contract](https://github.com/apache/lucene/blob/e04793d6517e17385b755f1707d59aeceaa41f15/lucene/core/src/java/org/apache/lucene/util/fst/Outputs.java#L28) for `Outputs` operations.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java (hunk 9 above)
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1210]: I think the index can be non-zero here. The last arc could contains suffix of the MSBVLong output.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java (hunk 11 above)
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnumFrame.java:109]: Hmm i'm a bit confused. I think `setFloorData` changes the position of `floorDataReader` because it calls
 `floorDataReader#reset`, so we can not move this line above?

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java (hunk 5 above)
Comment (gf2121) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/IntersectTermsEnumFrame.java:152]: I called it box because we are using a big `Long`... blockCode is better :)

**Comment by gf2121** [comment]
Thanks for review @mikemccand !

> but some head scratching -- hard to remember how these two crazy iterators work.

Agree that this is head scratching... I make a chart to try to explain this better.
&nbsp;
image

**Review by mikemccand — APPROVED** [review]
Looks great @gf2121 -- you want to merge and backport to 9.x?

**Comment by gf2121** [comment]
Thanks for review and great suggestions @mikemccand !

> you want to merge and backport to 9.x?

Yes. I'll merge and backport this this.

**Comment by gf2121** [comment]
Hi @mikemccand @jpountz !

The [newest nightly benchmark](https://home.apache.org/~mikemccand/lucenebench/2023.11.28.18.03.59.html) shows we get back some speed for [PKLookUp](https://home.apache.org/~mikemccand/lucenebench/PKLookup.html), but still not as good as the performance before the merge of  https://github.com/apache/lucene/pull/12631.

This result is not quite same as what i saw locally, but reasonable for me as we do more `Outputs#read` anyway. 

Since we are [preparing for the release of 9.9.0](https://lists.apache.org/thread/6bxf0fo1ypblxn2hbdfyg38y3hc8lo9g). I'm not sure if it is an acceptable trade-off now. Should we backport this or revert https://github.com/apache/lucene/pull/12631?  I'd appreciate your suggestions :)

**Comment by mikemccand** [comment]
I don't think we should revert #12631 -- that one was a nice disk space improvement, and the `PKLookup` performance is quite noisy -- going up or down by just as much when we upgrade JDKs or switch to Panama MMap impl.  Thanks @gf2121.

**Comment by jpountz** [comment]
+1 to not reverting #12631. The fact that our terms index used to not share prefixes properly was a bug, I'm glad it's fixed even though it may make performance slightly slower because more nodes have non-empty outputs to contribute.

FWIW it looks like your change did not only help PKLookup, but possibly also [`CountTerm`](https://home.apache.org/~mikemccand/lucenebench/CountTerm.html), [`Respell`](https://home.apache.org/~mikemccand/lucenebench/Respell.html), [`Fuzzy1`](https://home.apache.org/~mikemccand/lucenebench/Fuzzy1.html) and [`Fuzzy2`](https://home.apache.org/~mikemccand/lucenebench/Fuzzy2.html).

File: lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java (hunk 8 above)
Comment (benwtrent) [reply lucene/core/src/java/org/apache/lucene/codecs/lucene90/blocktree/SegmentTermsEnum.java:1189]: While we have strict contracts, I can see that `BytesSequenceOutputs#add(BytesRef, BytesRef)` has assertions that the length is > 0. 

Here in `OutputAccumulator` `readByte()` makes a big assumption that a `BytesRef` has at least length 1. If it had a length of 0, we would read past the ref end and read bytes sitting in a `byte[]` that we shouldn't.

IMO `OutputAccumulator` needs to be way more cautious than ``BytesSequenceOutputs#add(BytesRef, BytesRef)` because `OutputAccumulator` isn't making copies and is relying on the underlying byte arrays not changing.
